### PR TITLE
refactor: Replace const char* to std::string

### LINKED ISSUE #19000: Investigate where std::string should be returned instead of "const char*"
As noted by @MarcoFalke in https://github.com/bitcoin/bitcoin/pull/18994#discussion_r426250850 the case can be made that at least some of the functions in our code base returning `const char*` should return `std::string` instead.

Candidates:

```
$ git grep -E '^(static |inline |const |unsigned )*char *\*[^(<]*\(' ":(exclude)src/leveldb/" ":(exclude)src/univalue/" ":(exclude)src/serialize.h" ":(exclude)src/tinyformat.h"
src/bitcoin-cli.cpp:static const char *http_errorstring(int code)
src/script/script.cpp:const char* GetOpName(opcodetype opcode)
src/script/script.h:const char* GetOpName(opcodetype opcode);
src/script/script_error.cpp:const char* ScriptErrorString(const ScriptError serror)
src/script/script_error.h:const char* ScriptErrorString(const ScriptError error);
src/script/standard.cpp:const char* GetTxnOutputType(txnouttype t)
src/script/standard.h:const char* GetTxnOutputType(txnouttype t);
src/test/script_tests.cpp:static const char *FormatScriptError(ScriptError_t err)
```

#### Want to work on this issue?

The purpose of the `good first issue` label is to highlight which issues are suitable for a new contributor without a deep understanding of the codebase.

You do not need to request permission to start working on this. You are encouraged to comment on the issue if you are planning to work on it. This will help other contributors monitor which issues are actively being addressed and is also an effective way to request assistance if and when you need it.

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by kcalvinalvin**
Rationale: Addresses #19000  
Some functions should be returning std::string instead of const char*.
This commit changes that.

Main benefits/reasoning:

1.  The functions never return nullptr, so returning a string makes code at call sites easier to review (reviewers don't have to read the source code to verify that a nullptr is never returned)
2. All call sites convert to string anyway

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Concept ACK. It would be nice to motivate the change in the opening post a bit:

* The functions never return nullptr, so returning a string makes code at call sites easier to review (reviewers don't have to read the source code to verify that a nullptr is never returned)
* All call sites convert to string anyway

**Comment by kcalvinalvin** [comment]
> Concept ACK. It would be nice to motivate the change in the opening post a bit:
> 
>     * The functions never return nullptr, so returning a string makes code at call sites easier to review (reviewers don't have to read the source code to verify that a nullptr is never returned)
> 
>     * All call sites convert to string anyway

Thanks for the comment! Added those bits in the post

**Comment by MarcoFalke** [comment]
Also you can leave out the listing of files that you changed and what changes you made to those files. This is clear from reading the diff/looking at the changed files.

**Comment by kcalvinalvin** [comment]
> Also you can leave out the listing of files that you changed and what changes you made to those files. This is clear from reading the diff/looking at the changed files.

Ok. Should I take those out or just note for future PRs?

**Comment by MarcoFalke** [comment]
Both a suggestion for future PRs as well as this one, because the pull request description will end up in the merge commit, if this is merged.

**Comment by kcalvinalvin** [comment]
Done :)

**Review by practicalswift — COMMENTED** [review]
Concept ACK modulo nits

Welcome as a contributor @kcalvinalvin :)

**Comment by hebasto** [comment]
Concept ACK.

File: src/core_read.cpp [hunk 1]
```diff
@@ -34,14 +34,13 @@ CScript ParseScript(const std::string& s)
             if (op < OP_NOP && op != OP_RESERVED)
                 continue;
 
-            const char* name = GetOpName(static_cast<opcodetype>(op));
-            if (strcmp(name, "OP_UNKNOWN") == 0)
+            std::string name = GetOpName(static_cast<opcodetype>(op));
```
[the remark this answers is not in the document]
Comment (kcalvinalvin) [reply src/core_read.cpp]: Addressed in the new commit

File: src/core_read.cpp [hunk 2]
```diff
@@ -34,14 +34,13 @@ CScript ParseScript(const std::string& s)
...
+            if (name.compare("OP_UNKNOWN") == 0)
```
[the remark this answers is not in the document]
Comment (kcalvinalvin) [reply src/core_read.cpp]: Changed in the new commit

**Comment by kcalvinalvin** [comment]
Addressed ```GetTxnOutputType()``` in the new commit as requested [here](https://github.com/bitcoin/bitcoin/issues/19000#issuecomment-630205502)

File: src/core_write.cpp [hunk 3]
```diff
@@ -41,7 +41,7 @@ std::string FormatScript(const CScript& script)
                 ret += strprintf("%i ", op - OP_1NEGATE - 1);
                 continue;
             } else if (op >= OP_NOP && op <= OP_NOP10) {
-                std::string str(GetOpName(op));
+                std::string str = GetOpName(op);
```
Comment (hebasto) [inline src/core_write.cpp:44]: Why this change is needed?

File: src/script/standard.cpp [hunk 4]
```diff
@@ -25,7 +25,7 @@ WitnessV0ScriptHash::WitnessV0ScriptHash(const CScript& in)
     CSHA256().Write(in.data(), in.size()).Finalize(begin());
 }
 
-const char* GetTxnOutputType(txnouttype t)
+std::string GetTxnOutputType(txnouttype t)
```
Comment (hebasto) [inline src/script/standard.cpp:28]: This function has `return nullptr;` statement. It seems wrong :)

Comment (kcalvinalvin) [reply src/script/standard.cpp:28]: Yeah should be changed. Though I'm not sure it should be an assert or if it should return an Optional as stated [here](https://github.com/bitcoin/bitcoin/issues/19000#issuecomment-630228064)

File: src/core_write.cpp (hunk 3 above)
Comment (kcalvinalvin) [reply src/core_write.cpp:44]: I don't think it's needed. I'll revert the change.

File: src/script/standard.cpp (hunk 4 above)
Comment (MarcoFalke) [reply src/script/standard.cpp:28]: An assert seems appropriate, as this is unreachable code

Comment (hebasto) [reply src/script/standard.cpp:28]: [Developer Notes - C++ data structures](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#c-data-structures) could be useful :)

Comment (kcalvinalvin) [reply src/script/standard.cpp:28]: Changed to assert

**Comment by sipa** [comment]
It's unfortunate to make these functions return an `std::string` by value, as it implies a string copy every time they are called. This could be avoided by having a `static const std::string` inside the function for every value, and returning them as `const std::string&`.

**Comment by MarcoFalke** [comment]
All callers put the return value into a string anyway. So return value optimization (or whatever it is called) should take care of this and make it a no-op, no?

Going forward, with C++17 they can be made constexpr string views if performance is really a concern.

**Comment by tryphe** [comment]
Concept ACK

File: src/script/standard.h [hunk 5]
```diff
@@ -146,7 +146,7 @@ typedef boost::variant<CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash,
 bool IsValidDestination(const CTxDestination& dest);
 
 /** Get the name of a txnouttype as a C string, or nullptr if unknown. */
-const char* GetTxnOutputType(txnouttype t);
+std::string GetTxnOutputType(txnouttype t);
```
Comment (tryphe) [inline src/script/standard.h:149]: `#include <string>`

File: src/script/standard.cpp [hunk 6]
```diff
@@ -25,7 +25,7 @@ WitnessV0ScriptHash::WitnessV0ScriptHash(const CScript& in)
     CSHA256().Write(in.data(), in.size()).Finalize(begin());
 }
```
Comment (tryphe) [inline src/script/standard.cpp:27]: `#include <string>` :)

File: src/script/script.cpp [hunk 7]
```diff
@@ -7,7 +7,7 @@
 
 #include <util/strencodings.h>
 
-const char* GetOpName(opcodetype opcode)
+std::string GetOpName(opcodetype opcode)
```
Comment (tryphe) [inline src/script/script.cpp:10]: `#include <string>`

File: src/core_read.cpp [hunk 8]
```diff
@@ -34,10 +34,9 @@ CScript ParseScript(const std::string& s)
             if (op < OP_NOP && op != OP_RESERVED)
                 continue;
 
-            const char* name = GetOpName(static_cast<opcodetype>(op));
-            if (strcmp(name, "OP_UNKNOWN") == 0)
+            std::string strName = GetOpName(static_cast<opcodetype>(op));
```
Comment (tryphe) [inline src/core_read.cpp:37]: `#include <string>` (it's not included but should have been before anyhow)

File: src/bitcoin-cli.cpp [hunk 9]
```diff
@@ -157,7 +157,7 @@ struct HTTPReply
     std::string body;
 };
 
-static const char *http_errorstring(int code)
+static std::string http_errorstring(int code)
```
Comment (tryphe) [inline src/bitcoin-cli.cpp:160]: `#include <string>`

**Comment by sipa** [comment]
@MarcoFalke That's a good point - this PR isn't making anything worse in that case. It could be improved later, but no need add extra complexity for that now.

File: src/script/standard.h (hunk 5 above)
Comment (kcalvinalvin) [reply src/script/standard.h:149]: I've purposely only put ```#include <string>``` where the compiler yells at me since there are other ```.h``` and ```.cpp``` files that use ```std::string``` but omit ```#include <string>``` since there were some other file that had it.

Not quite sure which I should follow...

Comment (sipa) [reply src/script/standard.h:149]: You should include whatever you use in that module. There are certainly existing violations in the codebase, but that's not guidance.

Comment (kcalvinalvin) [reply src/script/standard.h:151]: Thanks for the input. Added #include <string>

File: src/script/standard.cpp (hunk 6 above)
Comment (kcalvinalvin) [reply src/script/standard.cpp:29]: addressed

File: src/script/script.cpp (hunk 7 above)
Comment (kcalvinalvin) [reply src/script/script.cpp:12]: addressed

File: src/core_read.cpp (hunk 8 above)
Comment (kcalvinalvin) [reply src/core_read.cpp:38]: addressed

File: src/bitcoin-cli.cpp (hunk 9 above)
Comment (kcalvinalvin) [reply src/bitcoin-cli.cpp:161]: addressed

**Review by hebasto — APPROVED** [review]
ACK bdd9cb6a9bc314400ad655a45ee32e03756468d1, compiled on Linux Mint 19.3 with GCC and Clang.

**Comment by MarcoFalke** [comment]
ACK bdd9cb6a9bc314400ad655a45ee32e03756468d1 🥑

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK bdd9cb6a9bc314400ad655a45ee32e03756468d1 🥑
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhkggv/dG77SPx7h89oenzVpqi337dajwuIwkr5AzDinyxCC+s+ZuWqRXM84r5Y
Kp2dppe4rvo5XnR6EYPgdGsICBt1tEarppmtEExnYrUfB7UB2PSyqfZ+A1juurnj
BTzRo019nO/RLavDVqEq/aqSa8fW8mkTu2gO95XLXsWFONojQAe0qRrMCwAAiviB
r54j6ApH2mJ+nllZqiI9yQxztUVFbIVWqZai+GIcy/wLMS6b6ssTMGH4y6NT1Ob4
O1t7eNENIZLv4RnEddCckA24rxVDOPE49ArD62E/yZn/9uzHfC+ZaFcCTyZ9eH12
h3bexZOOcnQes1nutjLYNVvG6+HAmaC3EjZuZhnX6wFvMUvrUzpCvbTe0WqRDy0u
GVER/1LkC0o3I+/xUADGzOutgKbviv8nqP1/fjdfkRXSbEMJ0KUknT6GYxbEADBu
NYf1BZdrK1hOmBQlO4N4upbt/UsnldngjW1xUPnCMtp9Ao+0le6qwzzZGC25qav6
vMt9IqV2
=535T
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `59bb0b7e1b6d4de434d0c1c688938d18335cc4075bc7302fd2f2754cdc746917  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010859bb0b7e1b6d4de434d0c1c688938d18335cc4075bc7302fd2f2754cdc746917f01064a023deb75484373a6ca919cad27cc608fff01016444d543267a04180868748dce0496008f1045ec51db5f008881db620b4eff75a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0107678ed8a5e763d01bd690e33b92e7a9e08f1045ec51db6f008d8fc2a80e0bb27f30083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010e2d3515eba371ceb51767db9a8ed610f08f1045ec51db5f0081226bf976b088ebb0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fc1717e10a8179aa608f963f817d51a108f1045ec51db5f008c03d4cb7f713b40e0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by laanwj** [comment]
ACK on the changes.

Can you please reformat your commit message though to put an empty line between the title and the body? Tools such as `git log --pretty=oneline` currently merge the entire description into the title, which is messy:
```
bdd9cb6a9bc314400ad655a45ee32e03756468d1 (pull/19004/head) script: Replace const char* to std::string Some functions should be returning std::string instead of const char*. This commit changes that.
```

**Comment by hebasto** [comment]
> Can you please reformat your commit message though to put an empty line between the title and the body?

... and replace `script: ` prefix with `refactor: ` ?

**Comment by kcalvinalvin** [comment]
> ACK on the changes.
> 
> Can you please reformat your commit message though to put an empty line between the title and the body?

Done :)

> ... and replace `script: ` prefix with `refactor: ` ?

Also Done :)

Also changed the GitHub PR title from `script:` prefix to `refactor:`

**Review by hebasto — APPROVED** [review]
re-ACK c57f03ce1741b38af448bec7b22ab9f8ac21f067

**Comment by MarcoFalke** [comment]
re-ACK c57f03ce17 (no changes since previous review) 🚃

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK c57f03ce17 (no changes since previous review) 🚃
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUg5Ugv/RLY5cmEYDu4MGRx80v7YSnQnpIxJ2l2+ls+NQLYL6wCj5oOG/g67mSvW
UCmzq51vj0FRhZJ+Kz54WUpHlizyWaettFcpytCug+O5u+7iWU5xwusrZTZxIIMq
hySm3NdHELk47dI2BTYAYtosJfEizyee/FimSovaWbokufmybwxXsjkMwz+NfeGv
D5qn0dh04Vs9y9GiN+9eSLMZn2b6tt7Wv5ryNjb+pNI7DzOfL2wiRqwGGRp+dAmO
cNcRzVxqmDXj+a23kiqMB7v9QujylftRsNnJ9gRXhNUDMxdGFRMG+tqcFKSWs/cU
3SI/n5y2ostgpVdFKaeKP56fiut0YUxKSrvlCZah3QIiEKMPiafILz1q8naN+Z5J
FxwIV0gpUBewkvBNv8Mmbi3c7bHAgNXHDr9XpWYq2N0wVgMjzv+bf2p55PlLi6HC
oU1IGMAMBSfLAzjxbhMQH1GBSdRX2NKqADGip4YlaXYiX4gYg8QQ/F9RPrVK1xZS
ohz+lirt
=GHzP
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `d852b864158d4b33654116fe600d925926dc1f39c2c521efb25067754e5f918d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108d852b864158d4b33654116fe600d925926dc1f39c2c521efb25067754e5f918df0108273dedcb477f825b0b9225598d0238c08fff010207c81b281a635d25f4d33fc58fa905e08f1045ec6b03bf008e3fe75a8ec53a3f20083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c579d590791e59ef9eef67c74c8dbb4608f1045ec6b03cf008d8a76626d4bcf44e0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010e11c29566cb7022ae8e3289f5b22b8f308f1045ec6b03bf008ea3e4c92f3a123910083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ecb6054a997b18ed16a7f5cbc0bd917708f1045ec6b03bf008a989cbbbdbf4aa5a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

**Comment by Empact** [comment]
~0 I like all of these except `GetOpName`, which could return `nullptr` in place of "OP_UNKNOWN" - seems reasonable to minimize processing in the unknown case:
https://github.com/bitcoin/bitcoin/compare/master...Empact:replace-char-with-string

**Comment by MarcoFalke** [comment]
Given that none of the call sites checked for nullptr of the char array, I'd find it very dangerous to re-introduce this sharp edge. Also for optional return values, we generally use Optional. In any case, I think this can be discussed in a follow-up pull request.

**Comment by Empact** [comment]
Fair enough, Code Review ACK https://github.com/bitcoin/bitcoin/pull/19004/commits/c57f03ce1741b38af448bec7b22ab9f8ac21f067

**Comment by practicalswift** [comment]
ACK c57f03ce1741b38af448bec7b22ab9f8ac21f067 -- patch looks correct
